static/js/message_list.js

```diff
@@ -384,6 +384,15 @@ export class MessageList {
     }
 
     update_topic_muting_and_rerender() {
+        // The logic inside `if` condition is necessary to collapse
+        // the muted topics (just to hide them from the viewer), if they
+        // mute the topic. This is done by filtering the unmuted messages
+        // and then we rerender the unmuted topics.
+        // But in the process, we don't rerender the newly muted topics.
+        // Hence, we rerender them by calling the rerender function
+        // for muted topics (which is done just before we update the
+        // items for topic muting).
+        this.rerender();
         this.data.update_items_for_topic_muting();
         if (this.data.excludes_muted_topics) {
             this.rerender();
```

Comment:
Why not just remove the condition? (But please make sure we rerender *after* updating the data)
I also did not understand your comment. Removing `this.rerender()` out of the condition is essential here because otherwise the bell icon will not update in stream+topic narrows, where `excludes_muted_topics` is set as `false`. Is there any other reason too?

Reply:
**Note:** This added logic works in a similar fashion to how we update the topic name in the recipient bar.

`rerender` function in `message_list.js` ultimately calls `render` function in `rerender_with_target_scrolltop` function inside `message_list_view.js`. So there you will observe that we render for `messages`(or `message_containers`) passed in the function and then we build `new_message_groups` there. So 2 things are important here the `messages` that we pass and the `new_message_groups` that are formed. Before we have called `this.data.update_items_for_topic_muting` which updates `this._items` (or `all_messages()`), it is necessary to `rerender` for the messages (which we have after muting/unmuting the topic). And `data.update_items_for_topic_muting` updates the `this._items` list (to store only `unmuted_messages`) to rerender them (or collapsing the view for muted topics).
After we have new message groups, it won't be possible to `populate_group_from_message_container` for the **original** message container (Sorry but this point will be a little complex to understand unless you look in the codebase).
`message_groups` differ for `current` and `home` `message_list`.

So, the 2nd call is important for the `unmuted_messages` to rerender them (or collapse the muted messages). Because we don't have them unless we call `data.update_items_for_topic_muting` and once we call `data.update_items_for_topic_muting` we don't have muted message in the `current`/`home` `message_list` and hence the first call is important in this way.
See the origin: https://github.com/zulip/zulip/commit/660475bd0cda520eea86a0e3bcfeedc5245ed352

I hope this clears.

Reply:
I understood your explanation, but not the reasoning behind why we need to care about the *original* (ie, pre-muting) messages/message groups/etc at all?
Once you update the set of messages in the `data`, the new groups will be created based on these new messages right?

It will be really helpful if you can mention what *user-facing* bug rerendering twice prevents.